source/jsonParsing.ml

```diff
@@ -7,60 +7,273 @@
+        | element ->
+            parse_value
+              element
+              (fun value -> parse_array (value :: current_values) post_parse_function)
+              decoder
+      and parse_object current_children post_parse_function decoder =
+        let location = Location.from_decoded_range (Jsonm.decoded_range decoder) in
+        match decode decoder with
+        | `Oe ->
+            post_parse_function
+              { Node.value = `Assoc (List.rev current_children); location }
+              decoder
+        | `Name name ->
+            parse_value
+              (decode decoder)
+              (fun value -> parse_object ((name, value) :: current_children) post_parse_function)
+              decoder
+        | _ ->
+            raise (ParseException { message = "Encountered unexpected token or element"; location })
+      in
+      let decoder = Jsonm.decoder (`String input) in
+      parse_value (decode decoder) (fun v _ -> v) decoder
+
+
+    let from_string input =
+      try Result.Ok (from_string_exn input) with
+      | ParseException { message; location } -> Result.Error { ParseError.message; location }
+
+
+    module Util = struct
+      let type_error expected json =
+        let message current = Format.sprintf "Expected %s, got %s" expected current in
+        let obj =
+          match json.Node.value with
+          | `Assoc _ -> "object"
+          | `Int _ -> "integer"
+          | `Float _ -> "float"
+          | `Null -> "null"
+          | `String _ -> "string"
+          | `Bool _ -> "boolean"
+          | `List _ -> "list"
+        in
+        TypeError { message = message obj; json }
+
+
+      let keys node =
+        match node.Node.value with
+        | `Assoc children -> List.map ~f:fst children
+        | _ -> []
+
+
+      (* Extract a child element with a particular key from a a json object. Raises TypeError for
+         other types. *)
+      let member key node =
+        match node.Node.value with
+        | `Assoc children -> (
+            match List.Assoc.find children key ~equal:String.equal with
+            | Some child -> child
+            | None -> null_node)
+        | _ -> raise (type_error "object" node)
+
+
+      let member_nullable key node =
+        match node.Node.value with
+        | `Assoc _ -> member key node
+        | _ -> null_node
+
+
+      let to_bool_option node =
+        match node.Node.value with
+        | `Bool b -> Some b
+        | _ -> None
+
+
+      let to_string node =
+        match node.Node.value with
+        | `String s -> s
+        | _ -> raise (type_error "string" node)
+
+
+      let to_int node =
+        match node.Node.value with
+        | `Int i -> i
+        | `Float f -> int_of_float f
+        | _ -> raise (type_error "integer" node)
+
+
+      let to_int_option node =
+        match node.Node.value with
+        | `Int _
+        | `Float _ ->
+            Some (to_int node)
+        | _ -> None
+
+
+      let to_list node =
```

Comment:
Suggestion: make all the functions that raise be suffixed with `_exn`, and have the functions returning `a' option`  not be suffix, to match the style of Core and how I have seen other code in our codebase.

Reply:
Great suggestion, wasn't aware :P - thanks!